README.md
# NAME

Device::PCA9685 - I2C interface to PCA9685 16 channel 12 bit PWM Generator

<div>
    <p>
    <img src="https://img.shields.io/badge/perl-5.10+-brightgreen.svg" alt="Requires Perl 5.10+" />
    <a href="https://travis-ci.org/shantanubhadoria/perl-Device-PCA9685"><img src="https://api.travis-ci.org/shantanubhadoria/perl-Device-PCA9685.svg?branch=build/master" alt="Travis status" /></a>
    <a href="http://matrix.cpantesters.org/?dist=Device-PCA9685%200.004"><img src="https://badgedepot.code301.com/badge/cpantesters/Device-PCA9685/0.004" alt="CPAN Testers result" /></a>
    <a href="http://cpants.cpanauthors.org/dist/Device-PCA9685-0.004"><img src="https://badgedepot.code301.com/badge/kwalitee/Device-PCA9685/0.004" alt="Distribution kwalitee" /></a>
    <a href="https://gratipay.com/shantanubhadoria"><img src="https://img.shields.io/gratipay/shantanubhadoria.svg" alt="Gratipay" /></a>
    </p>
</div>

# VERSION

version 0.004

# SYNOPSIS

        use Device::PWMGenerator::PCA9685;
    
        my $dev = Device::PWMGenerator::PCA9685->new(
            I2CBusDevicePath => '/dev/i2c-1', # this would be '//dev/i2c-dev-0 for Model A Pi
            debug            => 1,
            frequency        => 400, #Hz
        );
        $dev->enable();
        $dev->setChannelPWM(4,0,$dutycycle); # Duty cycle values between 0 and 4096 channel 4

# DESCRIPTION

This module provides a Object oriented interface to Adafruit 16 channel 12 bit
PWM Generator(PCA9685). Before you use this module make sure you follow install
Prerequisites for [Device::SMBus](https://metacpan.org/pod/Device::SMBus)

# ATTRIBUTES

## I2CBusDevicePath

this is the device file path for your I2CBus that the PCA9685 is connected on e.g. /dev/i2c-1
This must be provided during object creation.

## PWMGenerator 

    $self->PWMGenerator->enable();
    $self->PWMGenerator->setChannelPWM($channel,$pulseOnPoint,$pulseOffPoint);

The PCA9685 offers a 12 bit resolution which means across a duty cycle you may set and unset the PWM at 4096 different point. 
Range of values for $pulseOnPoint and $pulseOffPoint is 0 to 4095
Range of values of $channel is 0 to 15

This is a object of [Device::PWMGenerator::PCA9685](https://metacpan.org/pod/Device::PWMGenerator::PCA9685)

# SUPPORT

## Bugs / Feature Requests

Please report any bugs or feature requests through github at 
[https://github.com/shantanubhadoria/perl-device-pca9685/issues](https://github.com/shantanubhadoria/perl-device-pca9685/issues).
You will be notified automatically of any progress on your issue.

## Source Code

This is open source software.  The code repository is available for
public review and contribution under the terms of the license.

[https://github.com/shantanubhadoria/perl-device-pca9685](https://github.com/shantanubhadoria/perl-device-pca9685)

    git clone git://github.com/shantanubhadoria/perl-device-pca9685.git

# AUTHOR

Shantanu Bhadoria &lt;shantanu at cpan dott org>

# CONTRIBUTOR

Shantanu <shantanu@cpan.org>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2016 by Shantanu Bhadoria.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Changelog for Device-PCA9685

0.004 2016-04-07T06:20:13
 - Added travis (Shantanu Bhadoria)

0.003 2015-03-25T07:33:12
 - v0.003 (Shantanu Bhadoria)
 - fixed extends (Shantanu Bhadoria)

0.002 2013-12-15T02:37:02
 - v0.002 (Shantanu)
 - Fixed pod spellings (Shantanu)
 - Fixed POD (Shantanu)

0.001 2013-09-27T11:07:52
 - v0.001 (Shantanu)
 - Possibly first release (Shantanu)
 - Refactoring Submodule to allow easy loading of similar subdevices as
   plugins using a common prefix in MOP (Shantanu Bhadoria)
 - First Commit (Shantanu Bhadoria)


lib/Device/PCA9685/PWMGenerator.pm
package Device::PCA9685::PWMGenerator;

# PODNAME: Device::PCA9685::PWMGenerator
# ABSTRACT: I2C interface to PWM Generator on PCA9685 using Device::SMBus
# COPYRIGHT
# VERSION

# Dependencies
use 5.010;
use Moose;
use POSIX;

use Time::HiRes qw(usleep);

extends 'Device::SMBus';

use constant {
  MODE1              => 0x00,

  SUBADR1            => 0x02,
  SUBADR2            => 0x03,
  SUBADR3            => 0x04,

  PRESCALE           => 0xFE,

  LED0_ON_L          => 0x06,
  LED0_ON_H          => 0x07,
  LED0_OFF_L         => 0x08,
  LED0_OFF_H         => 0x09,

  ALLLED_ON_L        => 0xFA,
  ALLLED_ON_H        => 0xFB,
  ALLLED_OFF_L       => 0xFC,
  ALLLED_OFF_H       => 0xFD,
};

=attr I2CDeviceAddress

Containd the I2CDevice Address for the bus on which your PWM Generator is connected. It would look like 0x40. Default value is 0x40, If you have not set any of the six jumpers on the PCA9685, this should be the correct address and you will not have to modify this value in most cases.

=cut

has '+I2CDeviceAddress' => (
    is      => 'ro',
    default => 0x40,
);

=attr Frequency

Frequency of the PWM Pulse in Hz

=cut

has frequency => (
    is       => 'rw',
    required => 1,
    trigger  => \&_frequencySet,
);

has debug => (
    is      => 'ro',
    default => 1,
);

sub _frequencySet {
    my ( $self, $newFrequency, $oldFrequency ) = @_;

    # PCA9685 allows you to set the frequency using a register it calls PRESCALE
    say "Making PCA9685 go to sleep" if $self->debug;
    my $oldMode = $self->readByteData( MODE1 );
    my $newMode = ( $oldMode & 0x7f ) | 0x10; #set sleep bit while preserving other bits and ensuring that restart bit is set to 0 when writing to register
    $self->writeByteData( MODE1, $newMode );

    # Calculate Prescale from Frequency. Clock is 25 MHz on PCA8695
    my $prescale = floor( ( 25000000 / ( 4096 * $newFrequency ) ) + 0.5 ) - 1;
    say "Setting frequency to $newFrequency Hz. Calcualting prescale value \nprescale=round( 25000000/(4096 * frequency)) - 1 \nfrom frequency as $prescale(PCA9685 clock is at 25MHz or 25000000 Hz)" if $self->debug;
    if($prescale >= 0x03){
        $self->writeByteData( PRESCALE, $prescale );
    }else{
        die "Bad frequency value. Can't set prescale less than 3. reduce your frequency value";
    }
    $self->writeByteData( MODE1, $oldMode );
    usleep(5000);
    $self->writeByteData( MODE1, $oldMode | 0x80 ); # set extclk bit
}

=method enable

    $self->enable()

This function is just a placeholder. It is not required to call this function for this particular chipset

=cut

sub enable {
    my ( $self ) = @_;
    say "Setting up PCA9685" if $self->debug;
    $self->writeByteData( MODE1,0b00000000 );
}

=method setPWM

    $self->setPWM($channel,$pulseOnPoint,$pulseOffPoint)
    The PCA9685 offers a 12 bit resolution which means across a duty cycle you may set and unset the PWM at 4096 different point. 
    Range of values for $pulseOnPoint and $pulseOffPoint is 0 to 4095
    Range of values of $channel is 0 to 15

=cut

sub setChannelPWM {
    my ( $self, $channel, $on, $off ) = @_;

    say "Setting Channel $channel PWM on at $on step, off at $off step in 0 to 4095 steps at " . $self->frequency . "Hz" if $self->debug;
    $self->writeByteData( LED0_ON_L + ( 4 * $channel ), $on & 0xff );
    $self->writeByteData( LED0_ON_H + ( 4 * $channel ), $on >> 8 );

    $self->writeByteData( LED0_OFF_L + ( 4 * $channel ), $off & 0xff );
    $self->writeByteData( LED0_OFF_H + ( 4 * $channel ), $off >> 8 );
}

1;


lib/Device/PWMGenerator/PCA9685.pm
use strict;
use warnings;
package Device::PWMGenerator::PCA9685;

# PODNAME: Device::PWMGenerator::PCA9685
# ABSTRACT: I2C interface to PWM Generator on PCA9685 using Device::SMBus
# COPYRIGHT
# VERSION

# Dependencies
use 5.010;
use POSIX;

use Time::HiRes qw(usleep);

use Moose;
extends 'Device::SMBus';

=register MODE1

=register SUBADR1

=register SUBADR2

=register SUBADR3

=register PRESCALE

=register LED0_ON_L

=register LED0_ON_H

=register LED0_OFF_L

=register LED0_OFF_H

=register LED0_OFF_H

=cut

use constant {
  MODE1              => 0x00,

  SUBADR1            => 0x02,
  SUBADR2            => 0x03,
  SUBADR3            => 0x04,

  PRESCALE           => 0xFE,

  LED0_ON_L          => 0x06,
  LED0_ON_H          => 0x07,
  LED0_OFF_L         => 0x08,
  LED0_OFF_H         => 0x09,
};

=attr I2CDeviceAddress

Contains the I2CDevice Address for the bus on which your PWM Generator is connected. It would look like 0x40. Default value is 0x40, If you have not set any of the six jumpers on the PCA9685, this should be the correct address and you will not have to modify this value in most cases.

=cut

has '+I2CDeviceAddress' => (
    is      => 'ro',
    default => 0x40,
);

=attr Frequency

Frequency of the PWM Pulse in Hz

=cut

has frequency => (
    is       => 'rw',
    required => 1,
    trigger  => \&_frequencySet,
);

has debug => (
    is      => 'ro',
    default => 0,
);

sub _frequencySet {
    my ( $self, $newFrequency, $oldFrequency ) = @_;

    # PCA9685 allows you to set the frequency using a register it calls PRESCALE
    say "Making PCA9685 go to sleep" if $self->debug;
    my $oldMode = $self->readByteData( MODE1 );
    my $newMode = ( $oldMode & 0x7f ) | 0x10; #set sleep bit while preserving other bits and ensuring that restart bit is set to 0 when writing to register
    $self->writeByteData( MODE1, $newMode );

    # Calculate Prescale from Frequency. Clock is 25 MHz on PCA8695
    my $prescale = floor( ( 25000000 / ( 4096 * $newFrequency ) ) + 0.5 ) - 1;
    say "Setting frequency to $newFrequency Hz. Calcualting prescale value \nprescale=round( 25000000/(4096 * frequency)) - 1 \nfrom frequency as $prescale(PCA9685 clock is at 25MHz or 25000000 Hz)" if $self->debug;
    if($prescale >= 0x03){
        $self->writeByteData( PRESCALE, $prescale );
    }else{
        die "Bad frequency value. Can't set prescale less than 3. reduce your frequency value";
    }
    $self->writeByteData( MODE1, $oldMode );
    usleep(5000);
    $self->writeByteData( MODE1, $oldMode | 0x80 ); # set extclk bit
}

=method enable

    $self->enable()

This function is just a placeholder. It is not required to call this function for this particular chipset

=cut

sub enable {
    my ( $self ) = @_;
    say "Setting up PCA9685" if $self->debug;
    $self->writeByteData( MODE1,0b00000000 );
}

=method setChannelPWM

    $self->setChannelPWM($channel,$pulseOnPoint,$pulseOffPoint)
    The PCA9685 offers a 12 bit resolution which means across a duty cycle you may set and unset the PWM at 4096 different point. 
    Range of values for $pulseOnPoint and $pulseOffPoint is 0 to 4095
    Range of values of $channel is 0 to 15

=cut

sub setChannelPWM {
    my ( $self, $channel, $on, $off ) = @_;

    say "Setting Channel $channel PWM on at $on step, off at $off step in 0 to 4095 steps at " . $self->frequency . "Hz" if $self->debug;
    $self->writeByteData( LED0_ON_L + ( 4 * $channel ), $on & 0xff );
    $self->writeByteData( LED0_ON_H + ( 4 * $channel ), $on >> 8 );

    $self->writeByteData( LED0_OFF_L + ( 4 * $channel ), $off & 0xff );
    $self->writeByteData( LED0_OFF_H + ( 4 * $channel ), $off >> 8 );
}

1;
